File: pinyin-game-web/src/api/pinyinSelect.ts
import request, { type RequestConfig } from '@/utils/request'

export interface NextQuestion {
  question_id: number
  hanzi: string
  audio_url?: string | null
  index_no: number
  /** 本题无声母，可直接选韵母 */
  zero_initial?: boolean
}

export interface AnswerPayload {
  question_id: number
  initial: string
  final: string
  tone: number
  duration_ms: number
  session_id?: string
}

export interface AnswerResult {
  is_correct: boolean
  score_delta: number
  total_score: number
  correct_initial: string
  correct_final: string
  correct_tone: number
  pinyin_display: string
  hanzi: string
}

export interface PinyinStatistics {
  total_count: number
  correct_count: number
  accuracy: number
  total_score: number
}

export function fetchNextQuestion(params?: { session_id?: string; exclude_ids?: string }) {
  return request.get<any, NextQuestion>('/game/pinyin-select/question/next', { params })
}

export function submitPinyinAnswer(data: AnswerPayload) {
  return request.post<any, AnswerResult>('/game/pinyin-select/answer', data)
}

export function fetchPinyinStatistics(params?: { session_id?: string }) {
  return request.get<any, PinyinStatistics>('/game/pinyin-select/statistics', { params })
}

/** 声母/韵母格子朗读 */
export function fetchPartAudio(text: string, kind: 'initial' | 'final') {
  return request.get<any, { text: string; audio_url: string | null }>(
    '/game/pinyin-select/part-audio',
    { params: { text, kind }, silent: true } as RequestConfig,
  )
}

File: pinyin-game-api/app/schemas/pinyin_select.py
"""拼音练习游戏 Schema。"""

from typing import Optional

from pydantic import BaseModel, Field


class NextQuestionOut(BaseModel):
    """出题：不含正确答案。"""

    question_id: int
    hanzi: str
    audio_url: Optional[str] = None
    index_no: int = 1
    # 本题标准答案无声母时为 True，前台可直接选韵母（不展示「无」）
    zero_initial: bool = False


class AnswerIn(BaseModel):
    """提交用户选择的声母、韵母、声调。"""

    question_id: int
    initial: str = ""
    final: str
    tone: int = Field(ge=1, le=5)
    duration_ms: int = Field(ge=0, default=0)
    session_id: Optional[str] = None


class AnswerOut(BaseModel):
    """判题结果。"""

    is_correct: bool
    score_delta: int
    total_score: int
    correct_initial: str
    correct_final: str
    correct_tone: int
    pinyin_display: str
    hanzi: str


class PartAudioOut(BaseModel):
    """声母/韵母朗读。"""

    text: str
    audio_url: Optional[str] = None


class StatisticsOut(BaseModel):
    """答题统计。"""

    total_count: int
    correct_count: int
    accuracy: float
    total_score: int
